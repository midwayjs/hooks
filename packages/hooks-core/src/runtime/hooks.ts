import type { IMidwayContainer } from '@midwayjs/core'

import { ERR_INVALID_ARG_TYPE, validateString } from '../validator'
import { useContext } from './index'

export function useLogger() {
  const ctx = useContext()
  return ctx.logger
}

export function usePlugin(key: string): any {
  validateString(key, 'key')

  const ctx = useContext()
  return ctx.app[key] || ctx[key]
}

export function useInject<T = any>(identifier: new () => T): Promise<T>
export function useInject<T = any>(identifier: string): Promise<any>
export async function useInject(identifier) {
  if (typeof identifier !== 'function' && typeof identifier !== 'string') {
    throw new ERR_INVALID_ARG_TYPE('identifier', 'class, string', identifier)
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getAsync(identifier)
}

export function useConfig(key?: string) {
  if (key !== undefined) {
    validateString(key, 'key')
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getConfigService().getConfiguration(key)
}
