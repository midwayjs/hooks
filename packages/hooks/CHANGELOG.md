# @midwayjs/hooks

## 3.0.0-rc.17

### Patch Changes

- 468ffa4: fix: lazy load zod
- 711d651: feat: export HooksValidationError(#282)
- Updated dependencies [468ffa4]
- Updated dependencies [711d651]
  - @midwayjs/hooks-core@3.0.0-rc.17

## 3.0.0-rc.16

### Patch Changes

- 02c1e72: feat: add @midwayjs/hooks-upload component
- Updated dependencies [02c1e72]
  - @midwayjs/hooks-core@3.0.0-rc.16

## 3.0.0-rc.15

### Patch Changes

- f342e0a: @midwayjs/serve, fix path issue at windows, fixes #255
- Updated dependencies [f342e0a]
  - @midwayjs/hooks-core@3.0.0-rc.15

## 3.0.0-rc.14

### Patch Changes

- a11ce76: Fix class middleware support
- Updated dependencies [a11ce76]
  - @midwayjs/hooks-core@3.0.0-rc.14

## 3.0.0-rc.13

### Patch Changes

- 563430f: Hooks middleware now accept any types 589853fc001751f8ca94b2e04a0b5b416cdfe83f
- Updated dependencies [563430f]
  - @midwayjs/hooks-core@3.0.0-rc.13

## 3.0.0-rc.12

### Patch Changes

- e1feaea: Fix #239, use tuple to validate input args and update docs
- 3d6e603: Fix [#241](https://github.com/midwayjs/hooks/issues/241)
- a7e15f4: support spa for hooks-kit Fixed #238
- Updated dependencies [e1feaea]
- Updated dependencies [3d6e603]
- Updated dependencies [a7e15f4]
  - @midwayjs/hooks-core@3.0.0-rc.12

## 3.0.0-rc.11

### Patch Changes

- e991c98: use midway.config.js for api server project, fix #195
- 5883056: add @midwayjs/esrun to support decorator & tsconfig-paths, fix #207
- 30bc2f6: Add `@midwayjs/serve` to serve static files in FaaS environments, fix [#196](https://github.com/midwayjs/hooks/issues/196)
- e991c98: hooks-kit now support file-route
- Updated dependencies [e991c98]
- Updated dependencies [5883056]
- Updated dependencies [30bc2f6]
- Updated dependencies [e991c98]
  - @midwayjs/hooks-core@3.0.0-rc.11

## 3.0.0-rc.10

### Patch Changes

- 29d6d05: Add midwayjs deps to @midwayjs/hooks
- Updated dependencies [29d6d05]
  - @midwayjs/hooks-core@3.0.0-rc.10

## 3.0.0-rc.9

### Patch Changes

- 47b7842: add @midwayjs/hcc to compile project into single file
- 4a27b18: refactor: remove @midwayjs/hooks/test, import from @midwayjs/hooks directly
- 49f5113: Fix the duplicate path issue(#199)
- Updated dependencies [47b7842]
- Updated dependencies [4a27b18]
- Updated dependencies [49f5113]
  - @midwayjs/hooks-core@3.0.0-rc.9

## 3.0.0-rc.8

### Patch Changes

- Add support for serverless worker
- Updated dependencies
  - @midwayjs/hooks-core@3.0.0-rc.8

## 3.0.0-rc.7

### Patch Changes

- b15fda8: fix https://github.com/midwayjs/midway/issues/1641
- b15fda8: fix https://github.com/midwayjs/hooks/issues/193
- Updated dependencies [b15fda8]
- Updated dependencies [b15fda8]
  - @midwayjs/hooks-core@3.0.0-rc.7

## 3.0.0-rc.6

### Patch Changes

- Add missing deps
- Updated dependencies
  - @midwayjs/hooks-core@3.0.0-rc.6

## 3.0.0-rc.5

### Patch Changes

- Build
- Updated dependencies
  - @midwayjs/hooks-core@3.0.0-rc.5

## 3.0.0-rc.4

### Patch Changes

- Fix bugs
- Updated dependencies
  - @midwayjs/hooks-core@3.0.0-rc.4

## 3.0.0-rc.3

### Patch Changes

- Set correct version for examples
- Updated dependencies
  - @midwayjs/hooks-core@3.0.0-rc.3

## 3.0.0-rc.2

### Patch Changes

- Build after publish
- Updated dependencies
  - @midwayjs/hooks-core@3.0.0-rc.2

## 3.0.0-rc.1

### Patch Changes

- 71a3115: Release 3.0.0 rc version
- Updated dependencies [71a3115]
  - @midwayjs/hooks-core@3.0.0-rc.1
