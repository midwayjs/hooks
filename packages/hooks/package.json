{
  "name": "@midwayjs/hooks",
  "version": "3.0.1",
  "description": "A framework for building api & fullstack applications faster",
  "keywords": [
    "midway",
    "hooks",
    "fullstack",
    "api",
    "applications"
  ],
  "author": "Lxxyx <lxxyxzj@gmail.com>",
  "homepage": "https://github.com/midwayjs/hooks",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git@github.com:midwayjs/hooks.git"
  },
  "main": "dist/midwayjs-hooks.cjs.js",
  "module": "dist/midwayjs-hooks.esm.js",
  "browser": {
    "./dist/midwayjs-hooks.cjs.js": "./dist/midwayjs-hooks.browser.cjs.js",
    "./dist/midwayjs-hooks.esm.js": "./dist/midwayjs-hooks.browser.esm.js"
  },
  "dependencies": {
    "tslib": "^2.3.0",
    "lodash": "^4.17.21",
    "@midwayjs/hooks-internal": "^3.0.1",
    "@midwayjs/hooks-core": "^3.0.1",
    "@midwayjs/decorator": "^3.4.1",
    "@midwayjs/core": "^3.4.3"
  },
  "devDependencies": {
    "zod": "3.19.1",
    "@midwayjs/rpc": "3.0.1"
  },
  "peerDependencies": {
    "zod": ">=3.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist",
    "internal"
  ],
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "internal/index.ts"
    ]
  }
}
