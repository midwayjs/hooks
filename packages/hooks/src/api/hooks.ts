import type { IMidwayContainer } from '@midwayjs/core'
import {
  MidwayConfigService,
  getCurrentAsyncContextManager,
  ASYNC_CONTEXT_KEY,
} from '@midwayjs/core'
import { ERR_INVALID_ARG_TYPE, validateString } from '@midwayjs/hooks-core'

export function useContext<T = any>(): T {
  const ctx = getCurrentAsyncContextManager()
    .active()
    .getValue(ASYNC_CONTEXT_KEY)
  return ctx as T
}

export function useLogger() {
  const ctx = useContext()
  return ctx.logger
}

export function usePlugin(key: string): any {
  validateString(key, 'key')

  const ctx = useContext()
  return ctx.app[key] || ctx[key]
}

export function useInject<T = any>(identifier: new () => T): Promise<T>
export function useInject<T = any>(identifier: string): Promise<any>
export async function useInject(identifier) {
  if (typeof identifier !== 'function' && typeof identifier !== 'string') {
    throw new ERR_INVALID_ARG_TYPE('identifier', 'class, string', identifier)
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getAsync(identifier)
}

export function useConfig(key?: string) {
  if (key !== undefined) {
    validateString(key, 'key')
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.get(MidwayConfigService).getConfiguration(key)
}
