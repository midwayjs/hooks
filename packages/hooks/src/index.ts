// Runtime Api
export * from './api'
export type { ApiConfig } from '@midwayjs/hooks-core'

// User Config
export { defineConfig, UserConfig } from '@midwayjs/hooks-internal'

// For unit test
export type { BaseTrigger, HttpTrigger } from '@midwayjs/hooks-core'
export { getApiTrigger } from '@midwayjs/hooks-core'
