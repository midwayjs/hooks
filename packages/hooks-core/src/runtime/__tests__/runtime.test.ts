import { als, useContext } from '../'
import { useConfig, useInject, useLogger, usePlugin } from '../hooks'

const mockContext = {
  logger: jest.fn(),
  app: {
    mysql: jest.fn(),
  },
  cors: jest.fn(),
  requestContext: {
    getAsync: jest.fn(),
    getConfigService() {
      return {
        getConfiguration(key) {
          return key
        },
      }
    },
  },
}

test('test hooks runtime', async () => {
  await als.run({ ctx: mockContext }, async () => {
    expect(useContext()).toStrictEqual(mockContext)

    const logger = useLogger()
    logger()
    expect(mockContext.logger).toHaveBeenCalledTimes(1)

    const mysql = usePlugin('mysql')
    mysql()
    expect(mockContext.app.mysql).toHaveBeenCalledTimes(1)

    const cors = usePlugin('cors')
    cors()
    expect(mockContext.cors).toHaveBeenCalledTimes(1)

    await useInject('testService')
    expect(mockContext.requestContext.getAsync).toHaveBeenCalledTimes(1)

    const config = useConfig('test')
    expect(config).toStrictEqual('test')
  })
})

const cases = [null, 1, true, Symbol('test')]

test('usePlugin validate arguments', async () => {
  for (const cse of cases) {
    expect(() => (usePlugin as any)(cse)).toThrowErrorMatchingSnapshot()
  }
})

test('useConfig validate arguments', async () => {
  for (const cse of cases) {
    expect(() => (useConfig as any)(cse)).toThrowErrorMatchingSnapshot()
  }
})

test('useInject validate arguments', async () => {
  for (const cse of cases) {
    expect(() => (useInject as any)(cse)).rejects.toMatchSnapshot()
  }
})
