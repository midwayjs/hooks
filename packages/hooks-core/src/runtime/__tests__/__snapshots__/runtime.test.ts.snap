// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useConfig validate arguments 1`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received null"`;

exports[`useConfig validate arguments 2`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received type number (1)"`;

exports[`useConfig validate arguments 3`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received type boolean (true)"`;

exports[`useConfig validate arguments 4`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received type symbol (Symbol(test))"`;

exports[`useInject validate arguments 1`] = `[Error: [ERR_INVALID_ARG_TYPE]: The 'identifier' argument must be of type class, string. Received null]`;

exports[`useInject validate arguments 2`] = `[Error: [ERR_INVALID_ARG_TYPE]: The 'identifier' argument must be of type class, string. Received type number (1)]`;

exports[`useInject validate arguments 3`] = `[Error: [ERR_INVALID_ARG_TYPE]: The 'identifier' argument must be of type class, string. Received type boolean (true)]`;

exports[`useInject validate arguments 4`] = `[Error: [ERR_INVALID_ARG_TYPE]: The 'identifier' argument must be of type class, string. Received type symbol (Symbol(test))]`;

exports[`usePlugin validate arguments 1`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received null"`;

exports[`usePlugin validate arguments 2`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received type number (1)"`;

exports[`usePlugin validate arguments 3`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received type boolean (true)"`;

exports[`usePlugin validate arguments 4`] = `"[ERR_INVALID_ARG_TYPE]: The 'key' argument must be of type string. Received type symbol (Symbol(test))"`;
